### PR TITLE
xandikos: add tests and module

### PR DESCRIPTION
**Pull request by 0x4A6F**
###### Motivation for this change
Add tests and module for xandikos, which is included in #75532.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @matthiasbeyer

### CODE REVIEW

**Comment by 0x4A6F** [comment]
@GrahamcOfBorg test xandikos

**Comment by Ma27** [comment]
@GrahamcOfBorg test xandikos

File: nixos/modules/services/networking/xandikos.nix [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+      description = "A Simple Calendar and Contact Server";
+      after = [ "network.target" ];
+      wantedBy = [ "multi-user.target" ];
+
+      serviceConfig = {
+        User = "xandikos";
+        Group = "xandikos";
+        DynamicUser = "yes";
+        RuntimeDirectory = "xandikos";
+        StateDirectory = "xandikos";
```
Comment (Ma27) [inline nixos/modules/services/networking/xandikos.nix]: As this (AFAIK) lives in `/var/lib`: shouldn't this somehow depend on `cfg.directory`? (e.g. using `baseNameOf cfg.directory`)

**Review by Ma27 — COMMENTED** [review]
The change looks overall fine, thanks a lot!
Just a thought about the test: do we really need all of those VMs? I'm wondering as this means that you need notably more resources to execute the test-cases (although I'm not sure if we have some kind of "hard-limit" for tests, so feel free to disagree :D)

File: nixos/modules/services/networking/xandikos.nix (hunk 1 above)
Comment (Infinisil) [reply nixos/modules/services/networking/xandikos.nix]: Yeah this doesn't work like this. Imo the `cfg.directory` option should be removed.

File: nixos/modules/services/networking/xandikos.nix [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        StateDirectoryMode = "0700";
+        PrivateDevices = true;
+        ExecStart = "${cfg.package}/bin/xandikos --directory ${cfg.directory} --listen_address ${cfg.address} --port ${toString cfg.port} --route-prefix ${cfg.routePrefix} ${lib.concatStringsSep " " cfg.extraOptions}";
```
Comment (Infinisil) [inline nixos/modules/services/networking/xandikos.nix]: This line needs some line breaks!

File: nixos/modules/services/networking/xandikos.nix [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+          Extra command line arguments to pass to xandikos.
+        '';
+      };
+
+    };
+
+  };
+
+  config = mkIf cfg.enable {
+    environment.systemPackages = [ cfg.package ];
```
Comment (Infinisil) [inline nixos/modules/services/networking/xandikos.nix]: Does this need to be in the users environment? If not this can be removed

**Review by infinisil — CHANGES_REQUESTED** [review]

File: nixos/modules/services/networking/xandikos.nix (hunk 1 above)
Comment (0x4A6F) [reply nixos/modules/services/networking/xandikos.nix]: I'v droped cfg.directory and hard coded path.

File: nixos/modules/services/networking/xandikos.nix (hunk 2 above)
Comment (0x4A6F) [reply nixos/modules/services/networking/xandikos.nix]: Thanks, indeed. FIxed.

**Comment by 0x4A6F** [comment]
I've reduced the vm-count on nixos test.
But at least I'd like to cover the main two use-cases (standalone and behind proxy).

If there is no more improvements, I'd like to rebase commits.

Thanks for the reviews.

File: nixos/modules/services/networking/xandikos.nix [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+        defaultText = "pkgs.xandikos";
+        description = "The Xandikos package to use.";
+      };
+
+      address = mkOption {
+        type = types.str;
+        default = "localhost";
+        description = ''
+          The IP address on which Xandikos will listen.
+          If empty, listens on localhost.
```
Comment (aanderse) [inline nixos/modules/services/networking/xandikos.nix]: Is this true? Or should it say `By default listens on localhost`?

**Review by aanderse — COMMENTED** [review]
LGTM :+1:

@Ma27 @Infinisil any outstanding issues either of you see?

File: nixos/modules/services/networking/xandikos.nix (hunk 4 above)
Comment (0x4A6F) [reply nixos/modules/services/networking/xandikos.nix]: Should have been better worded "if not set". Changed to By default listens on localhost. Thanks.

**Review by infinisil — APPROVED** [review]

**Pull request merged**